mods/chrome-ui-tweaks.wh.cpp

```diff
@@ -0,0 +1,847 @@
+    if (code_section.size()==0) return 0;
+    Wh_Log(L"Searching for function %s", symbol_name);
+    const char* addr = unique_search(code_section, fsearch.search, symbol_name);
+    if (addr == NULL) {
+        Wh_Log(L"Could not find function %s; is the mod up to date?", symbol_name);
+        return NULL;
+    }
+    Wh_Log(L"Instructions were found at address: %p", addr);
+    int offset = fsearch.instr_offset;
+    const char* entry = addr - offset;
+    // verify the prologue is what we expect; otherwise search for it
+    // and verify it is preceded by 0xcc INT3 or 0xc3 RET (or ?? JMP)
+    auto prologue = fsearch.prologue;
+    if (prologue != std::string_view{entry, prologue.size()}) {
+        Wh_Log(L"Prologue not found where expected, searching...");
+        // maybe function length changed due to different compilation
+        auto search_space = std::string_view{entry - 0x40, addr};
+        size_t new_offset = search_space.rfind(prologue);
+        if (new_offset != std::string_view::npos) {
+            entry = (char*)search_space.begin() + new_offset;
+        } else {
+            entry = NULL;
+        }
+    }
+    if (entry) {
+        Wh_Log(L"Found entrypoint for function %s at addr %p", symbol_name, entry);
+        if (entry[-1]!=(char)0xcc && entry[-1]!=(char)0xc3) {
+            Wh_Log(L"Warn: prologue not preceded by INT3 or RET");
+        }
+        return entry;
+    } else {
+        Wh_Log(L"Err: Couldn't locate function entry point for symbol %s", symbol_name);
+        // log_hexdump(addr - 0x40, 0x5);
+        return NULL;
+    }
+}
+
+#pragma endregion "Code section scan for hooking without symbols"
+
+// ===========================================================================
+
+#pragma region "BookmarkBubble"
+// real type: void __thiscall (*)(ToolbarView* this, const GURL& url, bool)
+using ShowBookmarkBubble_t = void __thiscall (*)(void*, void*, bool);
+ShowBookmarkBubble_t ShowBookmarkBubble_Original;
+void __thiscall ShowBookmarkBubble_Hook(void* _this, void* url, bool already_bookmarked) {
+    // override already_bookmarked to disable "Simplified Bookmark Flow"
+    ShowBookmarkBubble_Original(_this, url, already_bookmarked || settings.bookmark_flow);
+}
+// setting up arguments to call ShowBubble from ToolbarView::ShowBookmarkBubble
+const std::string_view BookmarkBubble_instructions =
+    "\x88\x5c\x24\x30"sv        // mov byte ptr [rsp+0x30], bl  (already_bookmarked = bl)
+    "\x48\x89\x74\x24\x28"sv    // mov qword ptr [rsp+0x28], rsi
+    "\x48\x89\x6c\x24\x20"sv    // mov qword ptr [rsp+0x20], rbp
+    "\x48\x89\xf9"sv            // mov rcx, rdi
+    "\x48\x89\xc2"sv            // mov rdx, rax
+    "\x4d\x89\xf8"sv            // mov r8, r15
+    "\x4d\x89\xe1"sv            // mov r9, r12
+    "\xe8"sv;                   // call
+const std::string_view BookmarkBubble_prologue = "AWAVAUATVWUS"sv;
```

Comment:
[Nice one](https://stackoverflow.com/questions/39322552/meaning-of-a-common-string-in-executables) :)